### DISCUSSION TITLE
Embedding of a signature, which has been created by an API

### DISCUSSION THREAD

**Discussion by dimkar121** [MatthiasValvekens/pyHanko#235] [Q&A] [answered]
Hallo,
is there any way to embed a signature, which has been created by an API,  to a PDF file?

thanks in advance
Dimitrios

**Comment by MatthiasValvekens** [comment, marked as answer]
Yes, you can [implement your own `Signer` subclass](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#extending-signer).

**Reply by dimkar121** [reply]
Thanks Matthias!
I have extended the Signer class and it works fine for an atomic sign operation!
I have another issue now. 
I am gathering the digests of many documents, which are then sumbitted to an API, which in turn returns the signatures.
Is there any way to embed these signatures iteratively to the respective documents?
What is the optimal way to get the digests of many document and then sign these documents?
thanks in advance

**Reply by MatthiasValvekens** [reply]
Are we talking signatures for different documents? If so, you do that the same way you'd make anything else concurrent. There are multiple options.

 - Use [pyHanko's interrupted signing pattern](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#interrupted-signing) to "pause" the signing process after computing the document hash(es), then make whatever API calls you need to make, and call the appropriate finalization routine for each saved "intermediate" state.
 - If you're very comfortable with Python asyncio concurrency primitives, you can also use semaphores in your custom `Signer` to let coroutines know when their hashes have been signed. This works well enough when the number of documents is relatively small, and has the advantage that you can keep on using pyHanko's high-level API without having to dig too deep. The disadvantage is that it doesn't scale well in terms of resource management. PyHanko's [CSC signer](https://github.com/MatthiasValvekens/pyHanko/blob/master/pyhanko/sign/signers/csc_signer.py) uses something like this to implement batching.

Hope that helps.

**Reply by dimkar121** [reply]
Hi Matthias,
I have implemented the interrupted signing, but still the signature seems to get corrupted.
Should I put the DUMMY_HTTP_TS in finish_signing() and how (it throws an error, it cannot find this)?
Do I get the digest correctly?
```
digest = hashlib.sha256(prep_digest.document_digest).digest() 
```
I also attach the full snippet. The API expects a base64encoded digest and it reurns a raw base64encoded signature.

```

def instantiate_external_signer(sig_value: bytes):
      return signers.ExternalSigner(
            signing_cert=signing_cert,
            signature_value=sig_value,
            cert_registry=SimpleCertificateStore()
      )

async def prep_document(w):
    ext_signer = instantiate_external_signer(sig_value=bytes(256))
    pdf_signer = signers.PdfSigner(
        signers.PdfSignatureMetadata(
            field_name='SigNew',
            use_pades_lta=True,
            subfilter=fields.SigSeedSubFilter.PADES,
            md_algorithm='sha256'
        ),
        # note: this signer will not perform any cryptographic operations,
        # it's just there to bundle certificates with the generated CMS
        # object and to provide size estimates
        signer=ext_signer,
        timestamper=timestamps.HTTPTimeStamper('https://timestamp.aped.gov.gr/qtss/')
    )
    prep_digest, tbs_document, output = await pdf_signer.async_digest_doc_for_signing(w)
    md_algorithm = tbs_document.md_algorithm
    psi = tbs_document.post_sign_instructions
    signed_attrs = await ext_signer.signed_attrs(
        prep_digest.document_digest, 'sha256', use_pades=True
    )
    psi = tbs_document.post_sign_instructions
    return prep_digest, signed_attrs, psi, output




async def finish_signing(sig_value: bytes, prep_digest, signed_attrs,
                         psi, output_handle):
    # Here, assume sig_value is the signed digest of the signed_attrs
    # bytes, obtained from some remote signing service

    # use ExternalSigner to format the CMS given the signed value
    # we obtained from the remote signing service
    ext_signer = instantiate_external_signer(sig_value)
    sig_cms = await ext_signer.async_sign_prescribed_attributes(
        'sha256', signed_attrs=signed_attrs,
        timestamper=DUMMY_TS  #Should I put a dummy TS? 
    )

    await PdfTBSDocument.async_finish_signing(
        output_handle, prepared_digest=prep_digest,
        signature_cms=sig_cms,
        post_sign_instr=psi
    )




import hashlib
from pyhanko.pdf_utils.reader import PdfFileReader


async def full_procedure():
    # prepare document and signed attributes to sign
    with open("a.pdf", 'rb') as inf:
        w = IncrementalPdfFileWriter(inf, strict=False)
        prep_digest, signed_attrs, psi, output = await prep_document(w)
        print("digest before",prep_digest.document_digest)
        digest = hashlib.sha256(prep_digest.document_digest).digest()   # Am I getting the digest correctly?
        msg = base64.b64encode(digest)
        signature = getFromAPI(msg)
        signature = base64.b64decode(signature)
        await finish_signing(signature, prep_digest, signed_attrs, psi, output)
        with open("a_signed.pdf", "wb") as f:
            f.write(output.read())


asyncio.run(full_procedure())

```

Thanks in advance,
Dimitrios

**Reply by MatthiasValvekens** [reply]
Haven't looked in detail yet, but from a cursory inspection, I believe the thing you're doing wrong here is passing the document digest to the signing service. You need to have it sign the `signed_attrs` instead (possibly pre-hashed).

**Reply by dimkar121** [reply]
Hi Matthias,
Thanks for your reply!
This is what I'm sending to the API in the working version in async_sign_raw() of the Signer.
```
 hash_spec = get_pyca_cryptography_hash(
                    algos.SignedDigestAlgorithm(
                        {'algorithm': 'sha256_rsa'}
                    ).hash_algo
 )
 md = hashes.Hash(hash_spec)
 md.update(data)
 msg = md.finalize()
 msg = base64.b64encode(msg)
```
my best regards,
Dimitrios

**Reply by MatthiasValvekens** [reply]
Yes, but the `data` param in `async_sign_raw()` corresponds to `signed_attrs.dump()` in your new code. (If you just implement your own `Signer`, that's an irrelevant implementation detail; not so much in the interrupted signing workflow :) )

**Reply by dimkar121** [reply]
So, maybe, can this signed_attrs.dump()  do the job?

**Reply by dimkar121** [reply]
Yes Matthias, this  ```signed_attrs.dump()``` did the job!!!!
I needed to hash it exactly as I am doing in the atomic case.
Thanks for the support!!!!!
Kudos for the creation of this great package!!!

**Comment by deaw0003** [comment]
> Yes Matthias, this `signed_attrs.dump()` did the job!!!! I needed to hash it exactly as I am doing in the atomic case. Thanks for the support!!!!! Kudos for the creation of this great package!!!

Hi
 dimkar121. Can you please show the full snippet of your code that success in this scenario for learning purpose.
